### ISSUE TITLE
Use facts about unsigned 0 when checking ranges.

### ISSUE THREAD

**Issue by dtarditi** [microsoft/checkedc-clang#599]
Unsigned 0  <= any other unsigned integer.  We can use this fact when checking ranges during bounds declaration checking.  Given an array_ptr variable base and any unsigned integer expression e,  (base + 0) < (base + e).   This is assuming that array_ptrs have overflow checking on pointer arithmetic.

Using this fact would  let us eliminate warning messages when calling functions that take _Nt_array_ptrs with no declared bounds with  _Nt_array_ptr expressions that have bounds.  The default bounds for the arguments will be count(0), leading to the ranges mentioned above:
```
void f(nt_array_ptr<char> p) { …}

void g(nt_array_ptr<char> q : count(5)) {
  f(q);   // compiler warns that it can't prove that the argument has the correct bounds.
}
```

**Comment by ppashakhanloo** [comment]
I am trying to reproduce the warning. I have written this piece of code but it does not generate any warnings.

```
#include <stdchecked.h>

#pragma CHECKED_SCOPE ON

void f(nt_array_ptr<char> p) {
  nt_array_ptr<char> temp : count(5) = 0;
  p = temp;
}
void g(nt_array_ptr<char> q : count(5)) {
  f(q); // warning?
}
int main(int argc, nt_array_ptr<char> argv checked[] : count(argc)) {
  nt_array_ptr<char> a : count(5) = 0;
  g(a);
  return 0;
}
```
I am running checked-c Clang with no flags.
Should I run it with any particular flag to see the warning?

**Comment by dtarditi** [comment]
Try making the bounds for g a variable length.   When the bounds have the form of a constant-sized range, the compiler can make a decision.  The compiler doesn't understand that any expression e is >= 0 for unsigned integers. 

You can look at line 670 in https://github.com/microsoft/checkedc-parson/blob/master/parson.c to see a real example of the problem.

**Comment by dtarditi** [comment]
Completed by PR #624.
